examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+    "        [processed_dpo_dataset[x].prompt_mask for x in range(len(processed_dpo_dataset))],\n",
+    "        [processed_dpo_dataset[x].chosen_ids for x in range(len(processed_dpo_dataset))],\n",
+    "        [processed_dpo_dataset[x].chosen_mask for x in range(len(processed_dpo_dataset))],\n",
+    "        [processed_dpo_dataset[x].rejected_ids for x in range(len(processed_dpo_dataset))],\n",
+    "        [processed_dpo_dataset[x].rejected_mask for x in range(len(processed_dpo_dataset))],\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "R3SQTYTiwc-N"
+   },
+   "source": [
+    "# Define optimizer and DPO Trainer"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Optimizer, learning rate scheduler, gradient clipping\n",
+    "optimizer = optax.adamw(\n",
+    "    learning_rate=optax.schedules.warmup_cosine_decay_schedule(\n",
+    "        init_value=0.0,\n",
+    "        peak_value=LEARNING_RATE,\n",
+    "        warmup_steps=WARMUP_STEPS,\n",
+    "        decay_steps=MAX_STEPS,\n",
+    "        end_value=0.0,\n",
+    "    ),\n",
+    "    b1=B1,\n",
+    "    b2=B2,\n",
+    "    weight_decay=WEIGHT_DECAY,\n",
+    ")\n",
+    "if MAX_GRAD_NORM is not None:\n",
+    "  optimizer = optax.chain(\n",
+    "      optax.clip_by_global_norm(max_norm=MAX_GRAD_NORM),\n",
+    "      optimizer,\n",
+    "  )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Configure DPO Training (using previously defined config variables)\n",
+    "dpo_config = DpoTrainingConfig(\n",
+    "    beta=BETA,\n",
+    "    eval_every_n_steps=EVAL_EVERY_N_STEPS,\n",
+    "    max_steps=MAX_STEPS,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dpo_config"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dpo_trainer = DpoTrainer(\n",
+    "    model=lora_gemma,\n",
+    "    ref_model=gemma3,\n",
+    "    optimizer=optimizer,\n",
+    "    training_config=dpo_config,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "BdRa0cQHwgVO"
+   },
+   "source": [
+    "# Train and evaluate LoRA model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "show_hbm_usage()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Start training\n",
+    "print(\"Starting DPO training...\")\n",
+    "\n",
+    "dpo_trainer.train(train_ds=processed_dpo_dataset)\n",
+    "print(\"DPO training finished.\")"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
Markdown cell with heading: "# Evaluate". Can mention rough numbers, to show what kind of improvement we got